sys/include/wait.h

```diff
@@ -0,0 +1,234 @@
+ * add a strict dependency and let the application developer decide.
+ *
+ * @{
+ *
+ * @file
+ * @brief       Convenience functions to wait program execution
+ *
+ * @author      Marian Buschsieweke <marian.buschsieweke@posteo.net>
+ */
+
+#ifndef WAIT_H
+#define WAIT_H
+
+#include <assert.h>
+#include <stdint.h>
+
+#include "clk.h"
+#include "compiler_hints.h"
+#include "time_units.h"
+#include "wait_arch.h"
+#include "ztimer.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#if !defined(CPU_CYCLES_PER_WAIT_LOOP_ITERATION) || defined(DOXYGEN)
+/**
+ * @brief   Number of CPU cycles spent for each iteration of the wait loop
+ *
+ * @note    This can be fine-tuned per arch to get better accuracy. However, it
+ *          *must not* be overestimated: Overshooting (waiting longer than
+ *          asked for) is fine, but not the other way around.
+ *
+ * @warning Beware modern CPU architectures: On Cortex M7 CPUs with dual issue
+ *          a two instruction delay loop indeed takes only one cycle per
+ *          iteration once the dynamic branch prediction converges.
+ */
+#define CPU_CYCLES_PER_WAIT_LOOP_ITERATION 1
+#endif
+
+/**
+ * @name    Low accuracy "whatever is available" waits
+ *
+ * This group of functions wait execution with a low accuracy and use whatever
+ * backend is available. They do prefer low power variants if available anyway,
+ * but it is up the application developer to actually enable such backends
+ * (e.g. `ztimer_msec`) - if they are supported.
+ * @{
+ */
+
+/**
+ * @brief   Internal function to do busy wait using a down-counting loop
+ * @internal
+ *
+ * @param[in]   iterations  Number of loop iterations to wait
+ */
+static inline void wait_busy_loop(uint32_t iterations)
+{
+    /* The loop will not terminate if iterations is non-zero. We don't care
+     * about an overflow here, as `wait_busy_loop(UINT32_MAX)` is an insane
+     * thing to do. */
+    iterations++;
+    do {
+        iterations--;
+        /* tell compiler that iterations have been changed via inline asm
+         * constraints. No asm is actually needed */
+        __asm__ volatile(
+                ""
+                : "+r"(iterations)
+                : /* no inputs */
+                : /* no clobbers */
+        );
+    } while (iterations);
+}
+
+/**
+ * @brief   Delay program execution for roughly the given amount of
+ *          milliseconds
+ *
+ * @param[in]   msecs       Duration to wait execution in milliseconds
+ *
+ * @warning     This function gives **no guarantees** regarding accuracy or
+ *              power saving. Use @ref ztimer_sleep instead when you need
+ *              accuracy or power saving.
+ */
+static inline void wait_at_least_ms(uint16_t msecs)
+{
+    if (IS_USED(MODULE_ZTIMER_MSEC)) {
+        ztimer_sleep(ZTIMER_MSEC, msecs);
+    }
+    else {
+        uint32_t iterations = msecs * (uint64_t)coreclk() / MS_PER_SEC + 1;
+        wait_busy_loop(iterations);
+    }
```

Comment:
Are you sure that's better than a fallback to `ZTIMER_USEC`?
And why no `wait_at_least_us()`?

Reply:
> Are you sure that's better than a fallback to ZTIMER_USEC?

No, quite the opposite: I think falling back to `ZTIMER_USEC` is indeed better before falling back to the busy loop.

But I'd rather not have this PR being deadlocked. Especially since the DHT driver could profit from this getting upstream soonish.

Let's discuss this in the next sprint meeting. Maybe some agreement can be reached.

> And why no wait_at_least_us()?

That would not be allowed to fall back to `ZTIMER_MSEC`. But now that there is a CPU loop to fall back to, it would be indeed more than an alias for `ztimer_sleep(ZTIMER_USEC, ...)`. I'll add this again.